### ISSUE TITLE
Is that possible to io.write to console?

### ISSUE THREAD

**Issue by puttin** [Hammerspoon/hammerspoon#1063]

**Comment by asmagill** [comment]
At present, no... the `io` library works by using UNIX/POSIX file descriptors and the console does not correspond to such.  At present, I do not know how much work would be necessary to add this support, so no eta, but I welcome anyone else's input if they have any thoughts and I will give it some thought myself and see if it is something we can consider adding in the future.

**Comment by asmagill** [comment]
@cmsj, I think it may be worth re-opening this, but I want your thoughts first...

I have a module I worked on a while ago that duplicates the `stdout` and `stderr` pipes and passes whatever is sent to them to a lua function.  I've resurrected it at https://github.com/asmagill/hammerspoon_asm/tree/master/consolepipe, fixed a couple of errors that stumped me before, though I admit that it still has some issues before it could be used, most of them outlined in the README.md file.

The main reason I wanted to bounce this off of you is that our current implementation of the `print` function also passes its arguments into the original lua `print` function... this means that if I use consolepipe to write everything sent to `stdout` to the console, a giant feed-back loop begins and Hammerspoon basically locks up.  Is this call to lua's built in primitive necessary or a "just in case" from way back when?  I suspect that it might affect some of the output when debugging with XCode, but I don't know yet.  If so, having `stdout` send everything to our `print` without this call to the original primitive *may* be sufficient to work around it... let me know if you want me to setup a branch to test this out -- this could be built in to Hammerspoon's basic application rather than a module like this.

I don't know if there are other places that would have to be changed as well (hopefully none in the actual lua source itself!) to make a module like this viable for sending `stdout` to the console, but finding that the problem was easier then I originally thought has made me rethink the viability of it.

**Comment by cmsj** [comment]
@asmagill I don't remember specifically why the `rawprint(...)` call was added, but it does have the nice effect that it makes console output also appear in Xcode's output console.

What sort of usecases do you have in mind for being able to capture stdout/stderr?

**Comment by asmagill** [comment]
For their capture, not a whole lot, it was originally an exercise in "wonder if I can" many months ago and this issue reminded me of it... 

However if we *remove* the `rawprint` command from the `print` definition at https://github.com/Hammerspoon/hammerspoon/blob/master/extensions/_coresetup/init.lua#L105, then capture stdout as part of the Hammerspoon applications startup and route it through a callback which then "prints" it to the console then any lua code which "outputs" using `io.output:write(...)` or `io.stdout:write(...)` will output to the console...

I know for a fact that this is what LuaRocks does so its basically useless from within the Hammerspoon console to invoke the LuaRocks functions directly from within Hammerspoon (I investigated this back in the Mjolnir days when trying to create a built in package manager).

I don't know how much other third party code might output this way, so I don't know how big of a deal this really is... @puttin, what was your use case for this?

I'm not really advocating for the module specifically here, rather what the module points out we can do with a little modification to the Hammerspoon application -- have the console be the *actual* `stdout` (and maybe `stderr`, though that probably would need a filter to keep log messages that really *should* go into the system logs out of the Hammerspoon Console) for the lua process Hammerspoon runs.  Note that this code works by duplicating the pipes -- stdout and stderr are still open and are interpreted by macOS the same way... we just get control of a copy.

As to how it would affect output in XCode... don't know... we might need a flag or wrapper.  I should probably at least use XCode occasionally, now that I have a better machine, I'm just used to doing everything from the command line, but I can make sure to make this is part of what I test should you think this is worth looking into deeper.

**Comment by asmagill** [comment]
@cmsj, I had forgotten about your use of console output from within XCode when I made [this change](https://github.com/Hammerspoon/hammerspoon/commit/7876ea2d1bfa3d4c8d40b89c1182d23cf708c897)... should I change it to look for a flag (say `hs._suppressConsoleSTDOUT` and only skip the line when it's set to true?  That way it's on by default for you but can be disabled when desired?

**Comment by cmsj** [comment]
@asmagill no it's fine, I can just look in the console. Let's keep the code simpler :)

**Comment by jdtsmith** [comment]
Did this change ever go through?  I am working on getting "hs" to run in Emacs Lua-mode, and `io.write()` is the normal output methodology for capturing output from the lua shell.

**Comment by asmagill** [comment]
No, this stalled. But I see no reason not to re-open it... I've run for a while with the built in rawprint suppressed, so I don't believe it breaks anything (and the only way to be certain is to have others test it on their systems as well).

I'm out of town for the next week and can't do much with Hammerspoon for now, but I'll look into integrating this before we close off version 1.0.
